mitodl/mitxpro
# Added unused coupon reminder alert

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #277

#### What's this PR do?
Adds an alert/notification on page load if a user has unredeemed coupons assigned to them

#### How should this be manually tested?
- Create a 100% off coupon in the coupon creation form in ecommerce admin
- Send yourself that coupon in the bulk enrollment form in in ecommerce admin
- Instead of clicking the enrollment link mailed to you, just load any page beside the checkout page and make sure you perform a full browser refresh

You should see an alert with a link to go to the checkout page with that coupon applied. Clicking that link should get rid of the alert, and completing the order should get rid of the alert on successive page loads.

#### Any background context you want to provide?
[Design](https://projects.invisionapp.com/d/main/default/#/console/17338252/363331806/preview)

- The alert should not show up on the checkout page.
- I changed our bundles to make sure that the notifications code would be loaded in the React app and Django apps. This allows the notification to show up in all pages. There is a bunch of code repetition now between the `App` and `HeaderApp` components. I'll file a tech debt ticket to address that
- It was decided that we're not worrying about permanently dismissing this alert. If a user dismisses the alert then reloads the page, they will see it again until they actually complete the order with that coupon.
- If the coupon for some coupon assignment expires, we won't want to keep the coupon assignment around since unredeemed coupon assignments cause our user API view to run an extra query. I will file a tech debt ticket for that too, but it will be low priority since that scenario will be very rare
- ~~I tried to test the `componentDidUpdate` logic in App.js/HeaderApp.js, but I was unable to get enzyme to trigger that lifecycle event at all. I'll keep looking at that and will create a ticket if I'm not successful~~ (figured it out)

#### Screenshots (if appropriate)
![ss 2019-06-06 at 11 57 19 ](https://user-images.githubusercontent.com/14932219/59048904-d85b4000-8854-11e9-973c-755840a75a68.png)

![ss 2019-06-06 at 11 57 37 ](https://user-images.githubusercontent.com/14932219/59048912-dbeec700-8854-11e9-8618-b3c571731743.png)

![ss 2019-06-06 at 12 00 18 ](https://user-images.githubusercontent.com/14932219/59048917-de512100-8854-11e9-8798-876e65d69c9a.png)

![ss 2019-06-06 at 12 26 12 ](https://user-images.githubusercontent.com/14932219/59049573-4b18eb00-8856-11e9-91d1-f495f07b5fed.png)
